pycarol/carolina.py

```diff
@@ -25,49 +28,62 @@ def __init__(self, carol):
         self.cds_staging_rejected_path = None
 
     def init_if_needed(self):
+        """
+        Checks if backend credentials / object is still valid. This method is thread safe by using lock.
+
+        Returns:
+            True if backend has been reinitialized, false if it is still valid.
+
+        """
         expired = False
-        if self.client:
-            # Check if the token is not expired for at least another minute.. we do a little margin to avoid time difference issues
-            if self.expires_at is None or datetime.utcnow() + timedelta(minutes=1) < self.expires_at:
-                return
-            else:
-                expired = True
 
-        if Carolina.token is None or expired or Carolina.token.get('tenant_name', '') != self.carol.tenant['mdmName'] \
-                or Carolina.token.get('app_name', '') != self.carol.app_name \
-                or (datetime.utcnow() + timedelta(minutes=1)
-                    >= datetime.fromtimestamp(Carolina.token.get('expirationTimestamp', 1)/1000.0)):
-            token = self.carol.call_api('v1/storage/storage/token', params={'carolAppName': self.carol.app_name})
-            token['tenant_name'] = self.carol.tenant['mdmName']
-            token['app_name'] = self.carol.app_name
-            Carolina.token = token
+        # locks all instance variables to prevent inconsistencies
+        with Carolina.backend_lock:
+            if self.client:
+                # Check if the token is not expired for at least another minute.. we do a little margin to avoid time difference issues
+                if self.expires_at is None or datetime.utcnow() + timedelta(minutes=1) < self.expires_at:
+                    return False
+                else:
+                    expired = True
+
+            if Carolina.token is None or expired or Carolina.token.get('tenant_name', '') != self.carol.tenant['mdmName'] \
+                    or Carolina.token.get('app_name', '') != self.carol.app_name \
+                    or (datetime.utcnow() + timedelta(minutes=1)
+                        >= datetime.fromtimestamp(Carolina.token.get('expirationTimestamp', 1)/1000.0)):
+                token = self.carol.call_api('v1/storage/storage/token', params={'carolAppName': self.carol.app_name})
+                token['tenant_name'] = self.carol.tenant['mdmName']
+                token['app_name'] = self.carol.app_name
+                Carolina.token = token
+
+            token = Carolina.token
+            self.engine = token['engine']
 
```

Comment:
from this line you can leave the lock. the rest does not need to be lock

Reply:
same comment as above